api/envoy/extensions/filters/listener/proxy_protocol/v3/proxy_protocol.proto

```diff
@@ -17,4 +17,8 @@ option (udpa.annotations.file_status).package_version_status = ACTIVE;
 message ProxyProtocol {
   option (udpa.annotations.versioning).previous_message_type =
       "envoy.config.filter.listener.proxy_protocol.v2.ProxyProtocol";
+
+  // Types of TLV extensions that will be extracted from Proxy Protocol (V2) Header and
+  // saved to dynamic metadata.
+  repeated uint32 tlv_types = 1;
```

Comment:
> The protocol declares the type as a uint8, so I think it makes sense to make this a uint8 also.

that was my thought as well when I wrote the first draft, however it seems proto don't support a one-byte type.

Reply:
> What happens if a type is specified here, but isn't present?
> 
> What if the value is a number, not a string?
> 
> I think we should consider an configuration closer to what the header-to-metadata filter has. Even if it isn't all implemented at this point, the configuration should be made in a way we can get there.

A particular type of TLV's value will only be emitted to the dynamic metadata if both of the following conditions were met:
 1. the proxy protocol head has that type of TLV included (depending on the down stream traffic)
2. the type is included in the "tlv_types" list by the envoy's administrator.
I think it's possible that a TLV is not present in a connection even the type is in the list. The server don't have much control on the down stream traffic anyway. In such a case the decision on how to handle the traffic is left to a later time, e.g. a http filter. but not at the proxy protocol filter.

The value of a TLV can be a number, string or even a struct. No matter what's the actual value type, the values length is specified in the 2nd and 3rd byte in the TLV. With the length, we copy the value to the metadata without looking into the actual value. The reason is that the proxy protocol filter is responsible to save the TLV value while the actual business logic will happen at a later time. For example we may need both the TLV value and other information such as request header to make a decision.

Thanks for point me to the header-to-metadata filter. It seems has a flexible config 
  "on_header_present": "{...}",
  "on_header_missing": "{...}",
that can handle different use cases such as insert new values if a value is not present. I am not exactly sure how to define and implement that feature at this time. would it be possible to not included them at this PR? 

If you prefer I can add them, but probably have to leave them as not implemented yet. For the interface proto I can follow header-to-metadata's example:
  // list of type to overwrite value pairs, if we found the type in the proxy protocol header we will emit the overwrite value (not the one from the actual header) to metadata 
  "on_header_present": "{...}", 

 // list of type to overwrite value pairs, if we cannot  found the type in the proxy protocol header we will emit the overwrite value to metadata, otherwise we will emit the value from the actual header
  "on_header_missing": "{...}",

Please let me know what you think.

Reply:
@ggreenway seems like that could work, but do you need that level of flexibility? I'm not an expert on proxy protocol, are the valid values standards defined or user defined?

Reply:
There are a few that are standards-defined, and then the rest can be user-defined.

Reply:
I updated the config as suggested. The proto_format.sh seems have removed some of the comments though. I also copied this line

> option (udpa.annotations.versioning).previous_message_type =
        "envoy.config.filter.http.header_to_metadata.v2.Config.Rule";

but I was wondering is it needed?